substreams/crates/tycho-substreams/src/block_storage.rs

```diff
@@ -0,0 +1,79 @@
+use std::collections::HashMap;
+
+use substreams_ethereum::pb::{eth, eth::v2::block::DetailLevel};
+
+use crate::{
+    models::{ContractSlot, StorageChanges, Transaction},
+    pb::tycho::evm::v1::TransactionStorageChanges,
+};
+
+#[allow(dead_code)]
+/// Helper function to extract all storage changes on a block.
+/// The raw block information collected is intended to be used by the DCI (Dynamic Contract Indexer)
+/// to extract and index relevant changes. This is specifically for dynamically identified contracts
+/// that the DCI has chosen to index. Note that core protocol data should still be properly
+/// integrated and indexed by the substreams package as per usual.
+///
+/// ## Panics
+/// Panics if the provided block is not an extended block model, as indicated by its detail level.
+///
+/// ## Warning
+/// ⚠️ This function *only* works if the **extended block model** is available,
+/// more [here](https://streamingfastio.medium.com/new-block-model-to-accelerate-chain-integration-9f65126e5425)
+fn get_block_storage_changes(block: &eth::v2::Block) -> Vec<TransactionStorageChanges> {
+    if block.detail_level != Into::<i32>::into(DetailLevel::DetaillevelExtended) {
+        panic!("Only extended blocks are supported");
+    }
+    let mut block_storage_changes = vec![];
+
+    for block_tx in block.transactions() {
+        let transaction: Transaction = block_tx.into();
+
+        let mut storage_changes = vec![];
+        for call in block_tx.calls.iter() {
+            // Filter out calls that are reverted
+            if !call.state_reverted {
+                for storage_change in call.storage_changes.iter() {
+                    storage_changes.push(storage_change.clone());
+                }
+            }
+        }
```

Comment:
I think using functional programming is cleaner here, like this `storage_changes` doesn't need to be `mut`

```suggestion
let storage_changes: Vec<_> = block_tx.calls
    .iter()
    .filter(|call| !call.state_reverted)
    .flat_map(|call| call.storage_changes.iter().cloned())
    .sorted_unstable_by_key(|change| change.ordinal)
    .collect();
```

Reply:
I believe if we push functional programming to the extreme, the same function can be implemented in a more memory efficient way like this 

```
fn get_block_storage_changes(block: &eth::v2::Block) -> Vec<TransactionStorageChanges> {
    if block.detail_level != Into::<i32>::into(DetailLevel::DetaillevelExtended) {
        panic!("Only extended blocks are supported");
    }
    let mut block_storage_changes = vec![];

    for block_tx in block.transactions() {
        let tx_storage_changes: Vec<StorageChanges> = block_tx
            .calls
            .iter()
            .filter(|call| !call.state_reverted)
            .flat_map(|call| call.storage_changes.iter())
            .sorted_unstable_by_key(|change| change.ordinal)
            .fold(HashMap::new(), |mut acc: HashMap<(Vec<u8>, Vec<u8>), ContractSlot>, change| {
                acc.entry((change.address.clone(), change.key.clone()))
                    .or_default()
                    .push(ContractSlot {
                        slot: change.key.clone(),
                        value: change.new_value.clone(),
                    });
                acc
            })
            .into_iter()
            .map(|((address, _), slots)| StorageChanges { address, slots })
            .collect();

        block_storage_changes.push(TransactionStorageChanges {
            tx: Some(block_tx.into()),
            storage_changes: tx_storage_changes,
        });
    }

    block_storage_changes
}
```

Reply:
I think this looks cleaner, but the suggestions I made above are technically more efficient/optimised 👀

Reply:
I was able to avoid cloning so much and applied sorting on smaller groups (sorted on changes per address instead of sorted all changes at once and then grouped by address later). Also got it down to 2 loops/iters and avoided using complex interim hashmaps (like the fold is doing).

Reply:
I think both codes are equivalent with clones and loops 👀

But true, you're sorting many times smaller chunks of data and I'm sorting once a bigger chunk. And I think this makes your solution slightly more efficient

Reply:
my god you guys, this is crazy! my brain is still not this rust formatted yet 😅 I think I prefer @louise-poole 's suggestions just because I can read it better 🙃